### PR TITLE
Make backtraces work on Redox, copying Unix implementation

### PR DESCRIPTION
**Pull request by ids1024**
The `backtrace/` directory here is the same as the Unix one, except for adding an implementation of `get_executable_filename`.

### CODE REVIEW

File: src/libstd/sys/redox/backtrace/printing/mod.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+pub use self::imp::{foreach_symbol_fileline, resolve_symname};
+
+#[cfg(any(target_os = "macos", target_os = "ios",
+          target_os = "emscripten"))]
+#[path = "dladdr.rs"]
```
Comment (NilSet) [inline src/libstd/sys/redox/backtrace/printing/mod.rs]: Redox is never macos, so no need for dladdr.rs

Comment (ids1024) [inline src/libstd/sys/redox/backtrace/printing/mod.rs]: True; I thought it might be best not to modify those files too much (I just changed `get_executable_filename` in `mod.rs`)  so they can be easily synced with the Unix version, but otherwise several things could be removed.

The duplicated code is somewhat annoying; I don't think it's avoidable though.

Comment (ids1024) [inline src/libstd/sys/redox/backtrace/printing/mod.rs]: I guess (multiple) entire files that are useless is probably a waste, so I've pushed a commit addressing this.

**Comment by arielb1** [comment]
ping @brson - be our rubber stamp

**Comment by jackpot51** [comment]
@brson @arielb1 @shepmaster Any updates on this?

**Comment by alexcrichton** [comment]
Thanks @ids1024! Out of curiosity, where'd the changes to `config.sub` come from? We ideally like to stick pretty close to upstream, but if it's pretty easy to re-apply the changes to upstream that shouldn't be too bad.

**Comment by ids1024** [comment]
I could change the `config.sub` modification to just the single line required for Redox, but I though it was best to just update it from GNU: http://git.savannah.gnu.org/gitweb/?p=config.git;a=blob;f=config.sub

Future releases of libbacktrace will probably have that fix.

**Comment by alexcrichton** [comment]
Ok I don't really know what config.sub is anyway but the commit message is at least clear enough to me of how to reproduce it!

In that case let's...

@bors: r+

**Comment by ids1024** [comment]
`config.sub` is used by autotools to get the canonical name of the system from the target triplet (trivial for most systems, but a few have several variations of their triplets that it needs to handle). It needs to be updated when new systems are added.

`config.guess` is similar; it guesses a target triplet based on uname output. That is only relevant for native compilation though.

TLDR: It's one of the weird and ugly things in autotools that seem awful, but does help with portability to various strange systems.

This update seems to add fuchsia as well, so if they decide to implement Rust backtraces they won't run into that (small) issue.
